pkg/apis/kudo/v1alpha1/instance_types.go

```diff
@@ -16,65 +16,372 @@ limitations under the License.
+	for k, v := range i.Status.PlanStatus {
+		if v.Name == planStatus.Name {
+			i.Status.PlanStatus[k] = *planStatus
+			i.Status.AggregatedStatus.Status = planStatus.Status
+			if planStatus.Status.IsTerminal() {
+				i.Status.AggregatedStatus.ActivePlanName = ""
+			}
+		}
+	}
+}
 
-// PhaseStateError there was an error deploying the application.
-const PhaseStateError PhaseState = "ERROR"
+const snapshotAnnotation = "kudo.dev/last-applied-instance-state"
 
-// PhaseStateSuspend Spec was triggered to stop this plan execution.
-const PhaseStateSuspend PhaseState = "SUSPEND"
+// SaveSnapshot stores the current spec of Instance into the snapshot annotation
+// this information is used when executing update/upgrade plans, this overrides any snapshot that existed before
+func (i *Instance) SaveSnapshot() error {
+	jsonBytes, err := json.Marshal(i.Spec)
+	if err != nil {
+		return err
+	}
+	if i.Annotations == nil {
+		i.Annotations = make(map[string]string)
+	}
+	i.Annotations[snapshotAnnotation] = string(jsonBytes)
+	return nil
+}
+
+func (i *Instance) getSnapshotedSpec() (*InstanceSpec, error) {
+	if i.Annotations != nil {
+		snapshot, ok := i.Annotations[snapshotAnnotation]
+		if ok {
+			var spec *InstanceSpec
+			err := json.Unmarshal([]byte(snapshot), &spec)
+			if err != nil {
+				return nil, err
+			}
+			return spec, nil
+		}
+	}
+	return nil, nil
+}
+
+// selectPlan returns nil if none of the plan exists, otherwise the first one in list that exists
+func selectPlan(possiblePlans []string, ov *OperatorVersion) *string {
+	for _, n := range possiblePlans {
+		if _, ok := ov.Spec.Plans[n]; ok {
+			return kudo.String(n)
+		}
+	}
+	return nil
+}
+
+// GetPlanToBeExecuted returns name of the plan that should be executed
+func (i *Instance) GetPlanToBeExecuted(ov *OperatorVersion) (*string, error) {
+	if i.GetPlanInProgress() != nil { // we're already running some plan
+		return nil, nil
+	}
+
+	// new instance, need to run deploy plan
+	if i.NoPlanEverExecuted() {
+		return kudo.String("deploy"), nil
+	}
+
+	// did the instance change so that we need to run deploy/upgrade/update plan?
+	instanceSnapshot, err := i.getSnapshotedSpec()
+	if err != nil {
+		return nil, err
+	}
+	if instanceSnapshot.OperatorVersion.Name != i.Spec.OperatorVersion.Name {
+		// this instance was upgraded to newer version
+		log.Printf("Instance: instance %s/%s was upgraded from %s to %s operatorversion", i.Namespace, i.Name, instanceSnapshot.OperatorVersion.Name, i.Spec.OperatorVersion.Name)
+		plan := selectPlan([]string{"upgrade", "update", "deploy"}, ov)
+		if plan == nil {
+			return nil, &InstanceError{fmt.Errorf("supposed to execute plan because instance %s/%s was upgraded but none of the deploy, upgrade, update plans found in linked operatorVersion", i.Namespace, i.Name), kudo.String("PlanNotFound")}
+		}
+		return plan, nil
+	}
+	if !reflect.DeepEqual(instanceSnapshot.Parameters, i.Spec.Parameters) {
```

Comment:
nit:
```suggestion
	// did instance parameters change, so that the corresponding plan has to be triggered?
	if !reflect.DeepEqual(instanceSnapshot.Parameters, i.Spec.Parameters) {
```